jetty-util/src/main/java/org/eclipse/jetty/util/DateCache.java

```diff
@@ -152,89 +114,88 @@ public TimeZone getTimeZone()
     }
 
     /**
      * Format a date according to our stored formatter.
-     * If it happens to be in the same second as the last formatNow
-     * call, then the format is reused.
+     * If it happens to be in the same second as the last
+     * formatNow call, then the format is reused.
      *
-     * @param inDate the date in milliseconds since unix epoch
-     * @return Formatted date
+     * @param inDate the date in milliseconds since unix epoch.
+     * @return Formatted date.
      */
     public String format(long inDate)
     {
-        long seconds = inDate / 1000;
-
-        Tick tick = _tick;
+        return formatTick(inDate)._string;
+    }
 
-        // Is this the cached time
-        if (tick == null || seconds != tick._seconds)
-        {
-            // It's a cache miss
-            return ZonedDateTime.ofInstant(Instant.ofEpochMilli(inDate), _zoneId).format(_tzFormat);
-        }
+    /**
+     * Format a date according to our stored formatter without using the cache.
+     *
+     * @param inDate the Date.
+     * @return Formatted date.
+     */
+    public String formatWithoutCache(Date inDate)
+    {
+        return ZonedDateTime.ofInstant(inDate.toInstant(), _zoneId).format(_tzFormat);
+    }
 
-        return tick._string;
+    /**
+     * Format a date according to our stored formatter without using the cache.
+     *
+     * @param inDate the date in milliseconds since unix epoch.
+     * @return Formatted date.
+     */
+    public String formatWithoutCache(long inDate)
+    {
+        return ZonedDateTime.ofInstant(Instant.ofEpochMilli(inDate), _zoneId).format(_tzFormat);
     }
 
     /**
      * Format a date according to our stored formatter.
      * The passed time is expected to be close to the current time, so it is
      * compared to the last value passed and if it is within the same second,
-     * the format is reused.  Otherwise a new cached format is created.
+     * the format is reused. Otherwise, a new cached format is created.
      *
      * @param now the milliseconds since unix epoch
      * @return Formatted date
      */
+    @Deprecated
     public String formatNow(long now)
     {
-        long seconds = now / 1000;
-
-        Tick tick = _tick;
-
-        // Is this the cached time
-        if (tick != null && tick._seconds == seconds)
-            return tick._string;
         return formatTick(now)._string;
     }
 
+    @Deprecated
     public String now()
     {
         return formatNow(System.currentTimeMillis());
     }
 
+    @Deprecated
     public Tick tick()
     {
         return formatTick(System.currentTimeMillis());
     }
 
-    protected Tick formatTick(long now)
+    protected Tick formatTick(long inDate)
     {
-        long seconds = now / 1000;
+        long seconds = inDate / 1000;
 
         Tick tick = _tick;
         // recheck the tick, to save multiple formats
         if (tick == null || tick._seconds != seconds)
         {
```

Comment:
I don't think this logic is correct.

If the saved `_tick` is for this second, and I get a call to `formatTick()` with a time that is 5 seconds before, then the code caches the 5 seconds before tick.

Something like this should be more correct:

```java
if (tick != null && tick.seconds == seconds)
    return tick;

if (tick == null || tick.seconds < seconds)
{
    // format, cache and return.
}

// Return on-the-fly format but don't cache it.
```

Reply:
After discussion, ok to keep the old logic.